Add NrgKick Gen2 (via Modbus TCP)

wobu opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/14786

https://www.nrgkick.com/wp-content/uploads/2024/07/local_api_docu_simulate.html


- Uses Modbus TCP
- Has a dynamic capability check for 1p3p  and mA
- Sponsorship Authorization check

wobu committed:
Added types and initial implementation for NrgKick Gen2 charger json API

wobu committed:
added nrgkick gen2 template and added to  README.md

wobu committed:
fixed nrgkick gen2 types

wobu committed:
fixed type for nrgkick gen2 Control.CurrentSet

wobu committed:
fixed watt hours conversion for nrgkick gen2

wobu committed:
fixed missing colons in nrgkick gen2 types

wobu committed:
fixed changing current for nrgkick gen2

wobu committed:
fixed GetPhases for nrgkick gen2

wobu committed:
updated status and error code handling for NRGKick gen2

wobu committed:
applied requested changes to nrggen2 charger implementation

- renamed to be API implementation agnostic
- removed pointers in types definition
- removed unused code
- fixed error handlings

wobu committed:
removed unused nrgkick gen2 types

wobu committed:
implemented MaxCurrentMillis for nrgkick gen2 charger

andig committed:
Update nrggen2.go

andig committed:
Update nrggen2.go

wobu committed:
switched nrgkick gen2 to modbus tcp

wobu committed:
changed default modbus device id for nrgkick gen2

wobu committed:
implemented detectable capabilities for NRGKick Gen2

wobu committed:
added sponsor authorization check for nrgkick gen2

wobu committed:
fixed nrgkick-gen2 template description to mention modbus tcp

wobu committed:
Merge branch 'master' into feature/nrgkick-gen2-charger-modbus

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+		return nil, err
+	}
+
+	err = nrg.Enable(false)
```
That's nothing we should do when starting!

wobu replied:
the 1p3p and mA capabilities are now configurable via yaml template

andig replied:
Das wäre die letzte Option. Lassen sich die Optionen wirklich nicht auslesen?

wobu replied:
Ich hab in den offiziellen API Beschreibungen leider keine Hinweise darüber gefunden. Siehe Link im Pull Request

andig replied:
Könnte man den Strom testen? Also auslesen, und falls nciht auf .1A korrekt mal um +.1A schreibenm dann schauen ob das wieder eingelesen werden kann? 

Phasen konfigurierbar ist nachvollziehbar, Strom schon arg unglücklich. 

Gibts denn einen Grund anzunehmen, dass nciht _alle_ gen2 das können? Evtl. mal den Support fragen? Leider scheint es, als würde die Arbeit mit diesen Boxen sehr mühsam bleiben :(

wobu replied:
oh geez... Ich habs eben getestet. Man kann .1A setzen auch wenn die Funktion via App deaktiviert ist. Es scheint also als würde es immer einstellbar sein, nur via App ist die Funktion dann deaktiviert. Phasenumschaltung ist aber grundsätzlich deaktiviert. Den Hinweis in der Schnittstellendokumentation findet man ebenso dass man es zuvor via App aktivieren muss.
Ich schmeiss .1A Cap wieder raus und aktivier es dauerhaft.

wobu replied:
hast du noch Empfehlungen für Tests?
ich nutze aktuell lediglich den `evcc charger` CLI Befehl. Damit kann ich leider keine .1A setzen. Via UI bin ich derzeit überfordert wie ich einen Test durchführen könnte

andig replied:
Probier gerne mal https://github.com/evcc-io/evcc/pull/15341

wobu replied:
funktioniert, danke :)

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+	if b, err := nrg.conn.ReadHoldingRegisters(nrgKickGen2MaxPhases, 1); err == nil {
+		if maxPhases := encoding.Uint16(b); maxPhases > 1 {
+			if currentPhases, err := nrg.GetPhases(); err == nil {
+				if err := nrg.phases1p3p(2); err == nil {
```
Same here. We can't just switch around, much less twice.

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+		case 32:
+			fallthrough
+		case 33:
+			return api.StatusE, fmt.Errorf("%d", error)
```
I'd omit StatusE since it isn't handled by evcc- use F instead.

wobu replied:
i'll change, but there are some charges which are using the StatusE. Just want to point out

andig replied:
That's true- and thankfully these don't appear in practice ;)

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+			return api.StatusF, fmt.Errorf("%d", error)
+		}
+	case 7:
+		return api.StatusNone, nil
```
What's this?

wobu replied:
According to the specs status 7 is "WAKEUP". Any recommendations how to handle this in evcc?

andig replied:
No idea. Temporary like wakeup of the box? Or wakeup of what?

premultiply replied:
```suggestion
		return api.StatusB, nil
```

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+		return false, err
+	}
+
+	return !cast.ToBool(encoding.Uint16(b)), nil
```
```suggestion
	return binary.BigEndian.Uint16(b) == 1, nil
```

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+// MaxCurrentMillis implements the api.ChargerEx interface
+func (nrg *NRGKickGen2) maxCurrentMillis(current float64) error {
+	if current < 6 {
+		return fmt.Errorf("allowed range: 6.0 - rated_current (16.0A / 32.0A)")
```
Use same error message as other chargers

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+		return 0, err
+	}
+
+	return float64(encoding.Uint16(b)) / 10, nil
```
Use binary

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,365 @@
+		return 0, err
+	}
+
+	return int(encoding.Uint16(b)), nil
```
use binary

andig commented on templates/definition/charger/nrgkick-gen2.yaml:
```diff
@@ -0,0 +1,18 @@
+template: nrgkick-gen2
```
Please name template type and file same as charger implementation for sake of consistency. Changing either way is fine.

wobu committed:
small review changes

- used binary packages where possible
- removed use of StatusE
- consolidated maxCurrentMillis error message

wobu committed:
consolidated naming of nrggen2 template

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,367 @@
+		switch error := binary.BigEndian.Uint16(b); error {
+		case 32:
+			fallthrough
+		case 33:
```
```suggestion
		case 32, 33:
```

andig replied:
Aber: eigentlich ist das jetzt alles Default ;)

andig replied:
ping ;)

wobu committed:
updated status handling of nrggen2 charger

- Wakeup status is now StatusB

wobu committed:
added static capibility configuration of nrgick gen2 charger

wobu committed:
Merge commit 'a2e36a2efb13a3ecadc38c66f227b1d11afa16b3' into feature/nrgkick-gen2-charger-modbus

wobu committed:
removed static configurable mARegulation capability for nrgkick gen2 charger

- it is supported by default

andig commented on templates/definition/charger/nrggen2.yaml:
```diff
@@ -0,0 +1,29 @@
+      generic: Gen2
+requirements:
+  evcc: ["sponsorship"]
+  description:
+    de: NRGkick Gen2 Ladeeinheit via Modbus TCP
+    en: NRGkick Gen2 charging unit via Modbus TCP
```
```suggestion
```

andig commented on charger/nrggen2.go:
```diff
@@ -0,0 +1,342 @@
+		if err != nil {
+			return api.StatusNone, err
+		}
+
+		switch error := binary.BigEndian.Uint16(b); error {
+		case 32, 33:
+			return api.StatusF, fmt.Errorf("%d", error)
+		default:
+			return api.StatusF, fmt.Errorf("%d", error)
+		}
```
```suggestion
		return api.StatusF, fmt.Errorf("%d", error)
```

andig committed:
Apply suggestions from code review

andig commented:
Klasse, vielen Dank! Ab morgen im Nightly verfügbar!

JornHermans commented:
Works great! Thnx!
